cmd/precise-code-intel-bundle-manager/internal/database/cache.go

```diff
@@ -0,0 +1,153 @@
+package database
+
+import (
+	"sync"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/inconshreveable/log15"
+	"github.com/sourcegraph/sourcegraph/cmd/precise-code-intel-bundle-manager/internal/types"
+)
+
+// DatabaseCache is an in-memory LRU cache of Database instances.
+type DatabaseCache struct {
+	cache *ristretto.Cache
+}
+
+type databaseCacheEntry struct {
+	filename string
+	db       *Database
+	wg       sync.WaitGroup // user ref count
+	once     sync.Once      // guards db.Close()
+}
+
+// close closes the cached database value after its refcount has dropped to zero.
+// The underlying database's close method is guaranteed to be invoked only once.
+func (entry *databaseCacheEntry) close() {
+	entry.once.Do(func() {
+		go func() {
+			entry.wg.Wait()
+
+			if err := entry.db.Close(); err != nil {
+				log15.Error("Failed to close database", "filename", entry.filename, "error", err)
```

Comment:
nit: we use both "error" and "err", but since it's Go, I prefer "err" more, but totally up to you.
```suggestion
				log15.Error("Failed to close database", "filename", entry.filename, "err", err)
```

Reply:
Updated everywhere.